source/extensions/quic_listeners/quiche/envoy_quic_connection.h

```diff
@@ -28,43 +26,41 @@ class EnvoyQuicConnection : public quic::QuicConnection,
   EnvoyQuicConnection(const quic::QuicConnectionId& server_connection_id,
                       quic::QuicSocketAddress initial_peer_address,
                       quic::QuicConnectionHelperInterface& helper,
-                      quic::QuicAlarmFactory& alarm_factory, quic::QuicPacketWriter& writer,
+                      quic::QuicAlarmFactory& alarm_factory, quic::QuicPacketWriter* writer,
                       bool owns_writer, quic::Perspective perspective,
                       const quic::ParsedQuicVersionVector& supported_versions,
-                      Network::ListenerConfig& listener_config,
-                      Server::ListenerStats& listener_stats);
+                      Network::ConnectionSocketPtr&& connection_socket);
 
-  ~EnvoyQuicConnection() override = default;
+  ~EnvoyQuicConnection() override;
 
   // Called by EnvoyQuicSession::setConnectionStats().
   void setConnectionStats(const Network::Connection::ConnectionStats& stats) {
     connection_stats_ = std::make_unique<Network::Connection::ConnectionStats>(stats);
   }
 
-  // quic::QuicConnection
-  // Overridden to retrieve filter chain with initialized self address.
-  bool OnPacketHeader(const quic::QuicPacketHeader& header) override;
-
   // Called in session Initialize().
   void setEnvoyConnection(Network::Connection& connection) { envoy_connection_ = &connection; }
 
   const Network::ConnectionSocketPtr& connectionSocket() const { return connection_socket_; }
 
+  // Needed for ENVOY_CONN_LOG.
+  uint64_t id() const;
+
 protected:
   Network::Connection::ConnectionStats& connectionStats() const { return *connection_stats_; }
 
+  Network::Connection& envoyConnection() const {
+    ASSERT(envoy_connection_ != nullptr);
+    return *envoy_connection_;
+  }
+
 private:
   // TODO(danzh): populate stats.
   std::unique_ptr<Network::Connection::ConnectionStats> connection_stats_;
-  // Only initialized after self address is known. Must not own the underlying
-  // socket because UDP socket is shared among all connections.
+  // Assigned upon construction. Constructed with empty local address if unknown
+  // by then.
   Network::ConnectionSocketPtr connection_socket_;
   Network::Connection* envoy_connection_{nullptr};
```

Comment:
Maybe worth commenting this is functionally a pointer to the EnvoyQuicServerSession.

I had a moment of confusion why we were closing "the" connection on teardown but if course it's functionally closing the session which is fine.